### PR TITLE
Add prod env

### PR DESCRIPTION
**Pull request by robbie-daniels**
## What does this PR do
* Adds support for prod and staging environments
* Opening this PR for testing the app against the staging environment (the current staging app will not work, because the `public.json` is different
* The first commit is a proposal, I think it is safer to default to staging than to prod
* **NB this will not work until the corresponding back end changes have also been merged**

## How should this PR be tested?
- [ ] Manually inspect the proposed changes, and confirm that you agree with them
- [ ] Try deploying something into staging, and then using it
- [ ] Confirm that existing apps don't break (e.g. text-2-speech)

## Checklist before requesting a review
- [ ] I have performed a self-review of my code

### CODE REVIEW

**Comment by robbie-daniels** [comment]
Not clear why, but it appears to be working again.  I was able to deploy something successfully, and the other endpoints seem to work as expected.

**Comment by desimmons** [comment]
I'm seeing "404 page not found" for the READMEs being displayed through the assets

**Comment by desimmons** [comment]
> I'm seeing "404 page not found" for the READMEs being displayed through the assets

OK, I've resolve this issue I was seeing.It was caused by the READMEs for the old assets needing to be moved into the relevant staging bucket

File: assets/public.json [hunk 1]
```diff
@@ -28,9 +28,9 @@
...
     "POST_RUN_INFERENCE_ROUTE": "run",
     "POST_DEPLOY_ASSET_ROUTE": "asset",
-    "ROOT_DEPLOY_URL": "https://kernml-3mmgxhct.uc.gateway.dev",
+    "ROOT_DEPLOY_URL": "https://staging-3mmgxhct.uc.gateway.dev",
     "ROOT_CONSUMER_URL": "https://kernml-consumer-3mmgxhct.uc.gateway.dev",
-    "ROOT_RUN_URL": "https://kernml-run-3mmgxhct.uc.gateway.dev",
+    "ROOT_RUN_URL": "https://dhali-staging-run-3mmgxhct.uc.gateway.dev",
     "DHALI_PUBLIC_ADDRESS": "rstbSTpPcyxMsiXwkBxS9tFTrg2JsDNxWk",
     "DHALI_DEPLOYMENT_COST_PER_CHUNK_DROPS": 3000,
     "DHALI_EARNINGS_PERCENTAGE_PER_INFERENCE": 20,
```
Comment (desimmons) [inline assets/public.json:36]: Given that we have the `.prod.` and `.staging` versions of this file, do we still need this file?

**Review by desimmons — COMMENTED** [review]
Tested this change using tiny-server. Appears to behave as expected.The only thing I'm unsure of is whether we need the original `public.json` anymore

**Comment by robbie-daniels** [comment]
It’s there so the app runs out of the box locally, I figured that would be less confusing On Sun, Apr 30, 2023 at 16:17, desimmons ***@***.***> wrote:  
@desimmons commented on this pull request.

Tested this change using tiny-server. Appears to behave as expected.The only thing I'm unsure of is whether we need the original public.json anymore

In assets/public.json:
>      "POST_DEPLOY_README_ROUTE": "readme",
     "POST_RUN_INFERENCE_ROUTE": "run",
     "POST_DEPLOY_ASSET_ROUTE": "asset",
-    "ROOT_DEPLOY_URL": "https://kernml-3mmgxhct.uc.gateway.dev",
+    "ROOT_DEPLOY_URL": "https://staging-3mmgxhct.uc.gateway.dev",
     "ROOT_CONSUMER_URL": "https://kernml-consumer-3mmgxhct.uc.gateway.dev",
-    "ROOT_RUN_URL": "https://kernml-run-3mmgxhct.uc.gateway.dev",
+    "ROOT_RUN_URL": "https://dhali-staging-run-3mmgxhct.uc.gateway.dev",
     "DHALI_PUBLIC_ADDRESS": "rstbSTpPcyxMsiXwkBxS9tFTrg2JsDNxWk",
     "DHALI_DEPLOYMENT_COST_PER_CHUNK_DROPS": 3000,
     "DHALI_EARNINGS_PERCENTAGE_PER_INFERENCE": 20,

Given that we have the .prod. and .staging versions of this file, do we still need this file?

File: .github/workflows/firebase-hosting-merge-prod.yml [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+          mv lib/firebase_options{.prod,}.dart
+          cp assets/public{.prod,}.json
+      - run: flutter build web --release --dart-define=DEPLOYMENT_ENVIRONMENT=prod
+      - uses: FirebaseExtended/action-hosting-deploy@v0
+        if: github.ref == 'refs/heads/main'
+        with:
+          repoToken: '${{ secrets.GITHUB_TOKEN }}'
+          firebaseServiceAccount: '${{ secrets.FIREBASE_SERVICE_ACCOUNT_DHALI_PROD }}'
+          channelId: live
+          projectId: dhali-prod
```
Comment (desimmons) [inline .github/workflows/firebase-hosting-merge-prod.yml:27]: Nit: Add new line

File: assets/public.json [hunk 3]
```diff
@@ -28,9 +28,9 @@
     "POST_DEPLOY_README_ROUTE": "readme",
     "POST_RUN_INFERENCE_ROUTE": "run",
     "POST_DEPLOY_ASSET_ROUTE": "asset",
-    "ROOT_DEPLOY_URL": "https://kernml-3mmgxhct.uc.gateway.dev",
+    "ROOT_DEPLOY_URL": "https://staging-3mmgxhct.uc.gateway.dev",
     "ROOT_CONSUMER_URL": "https://kernml-consumer-3mmgxhct.uc.gateway.dev",
```
Comment (desimmons) [inline assets/public.json:32]: Should this URL be updated too?

Comment (desimmons) [inline assets/public.json:32]: Should this URL be updated too?

**Review by desimmons — APPROVED** [review]

File: assets/public.json (hunk 3 above)
Comment (robbie-daniels) [reply assets/public.json:32]: This one's correct

Comment (robbie-daniels) [reply assets/public.json:32]: Ditto

File: assets/public.staging.json [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    "CURRENCY_KEYS": {
+        "CODE": "code", 
+        "SCALE": "scale"
+    },
+    "GET_READMES_ROUTE": "readme",
+    "POST_DEPLOY_README_ROUTE": "readme",
+    "POST_RUN_INFERENCE_ROUTE": "run",
+    "POST_DEPLOY_ASSET_ROUTE": "asset",
+    "ROOT_DEPLOY_URL": "https://staging-3mmgxhct.uc.gateway.dev",
+    "ROOT_CONSUMER_URL": "https://kernml-consumer-3mmgxhct.uc.gateway.dev",
```
[the remark this answers is not in the document]
Comment (robbie-daniels) [reply assets/public.staging.json:32]: Yeah

**Pull request merged**